pallets/staking/src/lib.rs

```diff
@@ -2249,12 +2226,17 @@ decl_module! {
         pub fn set_commission_cap(origin, new_cap: Perbill) {
             T::RequiredCommissionOrigin::ensure_origin(origin.clone())?;
 
-            let old_cap = Self::validator_commission_cap();
-            ensure!(old_cap != new_cap, Error::<T>::NoChange);
-            <ValidatorCommissionCap>::put(new_cap);
-            // Update the validator prefs as per the `new_cap`.
-            // if `prefs.commission` of validator is > `new_cap`, it sets commission = new_cap.
-            Self::update_validator_prefs(new_cap);
+            // Update the cap, assuming it changed, or error.
+            let old_cap = ValidatorCommissionCap::try_mutate(|cap| -> Result<_, DispatchError> {
+                ensure!(*cap != new_cap, Error::<T>::NoChange);
+                Ok(core::mem::replace(cap, new_cap))
+            })?;
+            // Update `commission` in each validator prefs to `min(comission, new_cap)`.
+            <Validators<T>>::translate(|_, mut prefs: ValidatorPrefs| {
+                prefs.commission = prefs.commission.min(new_cap);
+                Some(prefs)
+            });
```

Comment:
A performance improvement here.

Reply:
The perf improvement was `::translate` -- it avoids doing the allocation in favor of in-place mutation.